segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -511,6 +516,43 @@ public void mergeSegments(MergeSegments mergeSegments) {
                     });
     }
 
+    @Override
+    public void mergeSegmentsBatch(WireCommands.MergeSegmentsBatch mergeSegments) {
+        final String operation = "mergeSegmentsBatch";
+
+        List<String> sources = mergeSegments.getSources();
+        for (String s : sources) {
+            if (!verifyToken(s, mergeSegments.getRequestId(), mergeSegments.getDelegationToken(), operation)) {
+                return;
+            }
+        }
+
+        log.info(mergeSegments.getRequestId(), "Merging Segments batch {} ", mergeSegments);
+
+        Function<String, CompletableFuture<Long>> mergeFunction = source ->
+                Futures.handleCompose(segmentStore.mergeStreamSegment(mergeSegments.getTarget(),
+                        source, TIMEOUT), (r, e) -> {
+                    if (e != null) {
+                        if (Exceptions.unwrap(e) instanceof StreamSegmentMergedException) {
+                            log.info(mergeSegments.getRequestId(), "Stream segment is already merged '{}'.",
+                                    sources);
+                            return segmentStore.getStreamSegmentInfo(mergeSegments.getTarget(), TIMEOUT)
+                                    .thenApply(SegmentProperties::getLength);
+                        } else {
+                            throw new CompletionException(e);
+                        }
+                    } else {
+                        recordStatForTransaction(r, mergeSegments.getTarget());
+                        return CompletableFuture.completedFuture(r.getTargetSegmentLength());
+                    }
+                });
+        CompletableFuture.completedFuture(sources.stream().map(x -> mergeFunction.apply(x).join()).collect(toList()))
```

Comment:
Why do you wrap this into a CompletableFuture? I think you can do this:

```
val mergeFutures = sources.stream()....;
Futures.allOfWithResults(mergeFutures). thenAccept(mergeResults -> ....))

```

What do you think?

Reply:
My understanding is that `Futures.allOfWithResults` does not **guarantee** "in order" execution of futures passed to it, hence this change.
In any case this code is not relevant now...